### PR TITLE
basic whisper implementation

### PR DESCRIPTION
**Pull request by markmeson**
thank you for considering

---
This change is [Reviewable](https://reviewable.io/reviews/dev-osrose/osirose-new/69)

### CODE REVIEW

File: src/map/src/cmapclient.cpp [hunk 1]
```diff
@@ -191,6 +194,35 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+                  GetId());
+    return true;
+  }
+
+  uint16_t _charID = GetId();
+  std::string _sender_name = entity_.component<BasicInfo>()->name_; 
+	std::string _recipient_name = P->targetId();
+	_recipient_name.erase(_recipient_name.size()-1); // for some reason this string has an extra char
+	std::string _message = P->message();
+	char *_recipient_client = NULL;
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: This should be a CMapClient* instead of char*.

File: src/map/src/cmapclient.cpp [hunk 2]
```diff
@@ -191,6 +194,35 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+	entitySystem_->process<BasicInfo, SocketConnector>([this, &_sender_name, &_recipient_client, _recipient_name] (Entity entity) {
+		if(_recipient_name == entity.component<BasicInfo>()->name_)
+			_recipient_client = (char *)entity.component<SocketConnector>()->client_;
+		//if(entity.component<SocketConnector>()->client_ == this)
+		//	_sender_name = entity.component<BasicInfo>()->name_;
+	});
+  logger_->trace("{} ({}) is whispering '{}' to {}", _sender_name, _charID,  _message, _recipient_name);
+	if(_recipient_client) {
+		auto packet = makePacket<ePacketType::PAKWC_WHISPER_CHAT>(_message, _sender_name);
+  	CMapServer::SendPacket((CMapClient *)_recipient_client, CMapServer::eSendType::PLAYER, *packet);
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: you can call directly _recipient_client->Send() instead. It's easier to understand I think.

Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: It also allow you to avoid modifying the CMapServer::SendPacket() function.

File: src/map/src/cmapclient.cpp [hunk 3]
```diff
@@ -191,6 +194,35 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+    return true;
+  }
+
+  uint16_t _charID = GetId();
+  std::string _sender_name = entity_.component<BasicInfo>()->name_; 
+	std::string _recipient_name = P->targetId();
+	_recipient_name.erase(_recipient_name.size()-1); // for some reason this string has an extra char
+	std::string _message = P->message();
+	char *_recipient_client = NULL;
+	entitySystem_->process<BasicInfo, SocketConnector>([this, &_sender_name, &_recipient_client, _recipient_name] (Entity entity) {
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: why send this and _sender_name?

File: src/rosecommon/include/packets/srv_whisper.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  SrvWhisper(const std::string &chat = "", const std::string &senderId = "")
+      : CRosePacket(ePacketType::PAKWC_WHISPER_CHAT),
+        chat_(chat),
+				senderId_(senderId) {}
+
+ protected:
+  void pack() { *this << senderId_ << chat_; }
+
+ private:
+  std::string chat_;
```
Comment (L3nn0x) [inline src/rosecommon/include/packets/srv_whisper.h]: please use message_ instead of chat_.

File: src/map/src/cmapclient.cpp (hunk 1 above)
Comment (markmeson) [inline src/map/src/cmapclient.cpp]: correct :)

File: src/map/src/cmapclient.cpp (hunk 2 above)
Comment (markmeson) [inline src/map/src/cmapclient.cpp]: ok i see what you mean thanks

File: src/map/src/cmapclient.cpp [hunk 5]
```diff
@@ -191,6 +194,33 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+                  GetId());
+    return true;
+  }
+
+  uint16_t _charID = GetId();
+  std::string _sender_name = entity_.component<BasicInfo>()->name_; 
+	std::string _recipient_name = P->targetId();
+	_recipient_name.erase(_recipient_name.size()-1); // for some reason this string has an extra char
+	std::string _message = P->message();
+	CMapClient *_recipient_client = NULL;
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: please use nullptr instead of NULL

File: src/map/src/cmapclient.cpp [hunk 6]
```diff
@@ -191,6 +194,33 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+  if (login_state_ != eSTATE::LOGGEDIN) {
+    logger_->warn("Client {} is attempting to whisper before logging in.",
+                  GetId());
+    return true;
+  }
+
+  uint16_t _charID = GetId();
+  std::string _sender_name = entity_.component<BasicInfo>()->name_; 
+	std::string _recipient_name = P->targetId();
+	_recipient_name.erase(_recipient_name.size()-1); // for some reason this string has an extra char
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: Could you look into it?

File: src/map/src/cmapclient.cpp [hunk 7]
```diff
@@ -191,6 +194,33 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+	std::string _message = P->message();
+	CMapClient *_recipient_client = NULL;
+	entitySystem_->process<BasicInfo, SocketConnector>([&_recipient_client, _recipient_name] (Entity entity) {
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: _recipient_name can be passed as a reference

File: src/map/src/cmapclient.cpp [hunk 8]
```diff
@@ -191,6 +194,33 @@ bool CMapClient::ChatReply(std::unique_ptr<RoseCommon::CliChat> P) {
...
+		if(_recipient_name == entity.component<BasicInfo>()->name_)
+			_recipient_client = entity.component<SocketConnector>()->client_;
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: if you return false in the if and true outside of the if, it'll magically skip all the entities once it has found the correct one ;)

File: src/map/src/cmapclient.cpp (hunk 7 above)
Comment (markmeson) [inline src/map/src/cmapclient.cpp]: roger

File: src/map/src/cmapclient.cpp (hunk 6 above)
Comment (markmeson) [inline src/map/src/cmapclient.cpp]: sure.  is there a way to output hex to logger, given that cout doesn't seem to work ?

File: src/map/src/cmapclient.cpp [hunk 9]
```diff
@@ -208,17 +208,25 @@ bool CMapClient::ChatWhisper(std::unique_ptr<RoseCommon::CliWhisper> P) {
 	std::string _recipient_name = P->targetId();
 	_recipient_name.erase(_recipient_name.size()-1); // for some reason this string has an extra char
 	std::string _message = P->message();
-	CMapClient *_recipient_client = NULL;
+	CMapClient *_recipient_client = nullptr;
 	entitySystem_->process<BasicInfo, SocketConnector>([&_recipient_client, _recipient_name] (Entity entity) {
```
Comment (L3nn0x) [inline src/map/src/cmapclient.cpp]: it's not process anymore but processEntities

**Pull request merged**